presto-main/src/main/java/com/facebook/presto/operator/GroupIdOperator.java

```diff
@@ -0,0 +1,239 @@
+
+            // will have a 'true' for every channel that should be set to null for each grouping set
+            BitSet[] groupingSetNullChannels = new BitSet[groupingSetChannels.size()];
+            for (int i = 0; i < groupingSetChannels.size(); i++) {
+                groupingSetNullChannels[i] = new BitSet(inputTypes.size());
+                // first set all grouping columns to true
+                for (Integer groupingColumn : allGroupingColumns) {
+                    groupingSetNullChannels[i].set(groupingColumn, true);
+                }
+                // then set all the columns in this grouping set to false
+                for (Integer nonNullGroupingColumn : groupingSetChannels.get(i)) {
+                    groupingSetNullChannels[i].set(nonNullGroupingColumn, false);
+                }
+            }
+
+            return new GroupIdOperator(operatorContext, outputTypes, groupingSetNullChannels);
+        }
+
+        @Override
+        public void close()
+        {
+            closed = true;
+        }
+
+        @Override
+        public OperatorFactory duplicate()
+        {
+            return new GroupIdOperatorFactory(operatorId, inputTypes, groupingSetChannels);
+        }
+    }
+
+    private final OperatorContext operatorContext;
+    private final List<Type> types;
+    private GroupIdRowGenerator groupIdRowGenerator;
+    private boolean finishing;
+
+    public GroupIdOperator(
+            OperatorContext operatorContext,
+            List<Type> inputTypes,
+            BitSet[] groupingSetNullChannels)
+    {
+        this.operatorContext = requireNonNull(operatorContext, "operatorContext is null");
+        this.types = requireNonNull(inputTypes, "inputTypes is null");
+
+        requireNonNull(groupingSetNullChannels, "groupingSetChannels is null");
+        this.groupIdRowGenerator = new GroupIdRowGenerator(types, groupingSetNullChannels);
+    }
+
+    @Override
+    public OperatorContext getOperatorContext()
+    {
+        return operatorContext;
+    }
+
+    @Override
+    public List<Type> getTypes()
+    {
+        return types;
+    }
+
+    @Override
+    public void finish()
+    {
+        finishing = true;
+    }
+
+    @Override
+    public boolean isFinished()
+    {
+        return finishing && !groupIdRowGenerator.hasBufferedRows();
+    }
+
+    @Override
+    public boolean needsInput()
+    {
+        return !finishing && !groupIdRowGenerator.hasBufferedRows();
+    }
+
+    @Override
+    public void addInput(Page page)
+    {
+        checkState(!finishing, "Operator is already finishing");
+        requireNonNull(page, "page is null");
+
+        groupIdRowGenerator.addNewPage(page);
+    }
+
+    @Override
+    public Page getOutput()
+    {
+        if (!groupIdRowGenerator.hasBufferedRows()) {
+            return null;
+        }
+
+        return groupIdRowGenerator.nextPage();
+    }
+
+    private static class GroupIdRowGenerator
```

Comment:
At the beginning, I thought this is just an iterator. But then I realized `addNewPage` is called from the parent class. I find this a very weird abstraction (reusable iterator). Given the parent class is not doing anything, maybe this class should just be inlined.